metacat-main/src/main/java/com/netflix/metacat/main/services/MetadataService.java

```diff
@@ -102,4 +131,73 @@ public void processDeletedDataMetadata() {
         }
         log.info("End deleting data metadata");
     }
+
+    /**
+     * Deletes definition metadata of tables/views/partitions that have been deleted already.
+     */
+    public void deleteObsoleteDefinitionMetadata() {
+        log.info("Start deleting obsolete definition metadata");
+        final MetacatRequestContext metacatRequestContext = MetacatContextManager.getContext();
+        List<DefinitionMetadataDto> dtos = null;
+        int offset = 0;
+        final int limit = 10000;
+        final ThreadServiceManager threadServiceManager =
+            new ThreadServiceManager(registry, 25, 20000, "definition");
+        final ListeningExecutorService service = threadServiceManager.getExecutor();
+        while (offset == 0 || dtos.size() == limit) {
+            dtos = userMetadataService.searchDefinitionMetadatas(null, null, null,
+                "id", null, offset, limit);
+            dtos.forEach(dto ->
+                service.submit(() -> deleteDefinitionMetadata(dto.getName(), false, metacatRequestContext)));
+            offset += limit;
+        }
+        threadServiceManager.stop();
+        log.info("End deleting obsolete definition metadata");
```

Comment:
We should log out the number of deleted records. It will be helpful to bucket them into catalogs, but at least the count will be useful.